moves monster

JaskRendix opened the pull request:
PR addresses monsters and moves.
This change involves only the structure.

@Sanglorian 
Since the addition of the last monsters and the fix of weight and height, I was working on this PR.
I share this after the max_moves (as you said in another PR, max 4), a mechanism of overwriting moves that it's really intriguing.

Example: **Rockitten** > **Rockat** > **Jemuar**

Situation at the moment, 
**Rockitten** moves: ram, boulder and thunderball
**Rockat** moves: ~~ram~~, glover and ~~thunderball~~
**Jemuar** moves: ~~ram~~, ~~glover~~ and stampede
**Evolutions**: **Rockitten** > lv 24 > **Rockat** > lv 36 > **Jemuar**
`Total potential moves without duplicates: 5`

After the PR:
**Rockitten** moves: ram, boulder and thunderball
**Rockat** moves: gover, ice_claw, surge
**Jemuar** moves: stampede, mudslide, biting_winds
**Evolutions**: **Rockitten** > lv 24 > **Rockat** > lv 42 > **Jemuar** (changed because the monster learns 3 moves every 4 level, then evolves, it's explained inside the file attached)
`Total potential moves without duplicates: 9`

Confrontation **Jemuar** JSON:
before:
```
2 lv - ram
2 lv - glower
2 lv - stampede
```
after:
```
2 lv - ram
2 lv - boulder
12 lv - thunderball
28 lv - glower
32 lv - ice_claw
36 lv - surge
44 lv - stampede
48 lv - mudslide
52 lv - biting_winds
```
The mechanism:
- basic and standalone (no evo) monsters (eg. rockitten is basic) have 3 moves;
- 1 stage (rockat) have 6 moves;
- 2 stage (jemuar) have 9 moves;

I added also a new validator called multiple to check both **level_learned** for techniques and **at_level** for evolutions. Technique learning is multiple of 2, while evolution level is multiple of 3, the will is having two ways, so we can reduce drastically the probability that someone needs to overwrite the technique immediately after an evolution.

Using the script for the missing move (the other PR) I generated the missing techniques based on type, everything related the moves is here: [moves.ods](https://github.com/Tuxemon/Tuxemon/files/10298760/moves.ods)
Everything related to the change of "at_level" involving evolutions is here: [at_level evolution.ods](https://github.com/Tuxemon/Tuxemon/files/10298761/at_level.evolution.ods)
Both files contain all the changes and additions I made.

Do you want also standalone monster with more techniques? We can generate easily, it doesn't take much time.
Do you think we need to use the 9 moves list (jemuar) also for rockitten and rockat? At the end someone can also refuse the evolution and keeping developing rockitten until the max level.

Let me know.

I think in the second case, considering overwriting and the maximum number of moves (4), it can be more challenging.
I already tested some mechanism and these are really interesting.

Sanglorian commented:
This sounds like a great improvement. The 3 techniques all at the same time was just a place holder so I like your approach. The only thing I'd change is keeping the evolutions at level 36. Other evolution changes are fine if it's just a few levels difference.

For earlier evolutions, e.g. Rockitten, it'd be good if they got some techniques even after they would normally evolve, but not all of them.

JaskRendix commented:
No problem, I fixed the 4 cases you told me about (42 instead of 36), I changed also the level of the move, because I want to avoid a double trigger if I can.

I guess we can level up to 6 techniques for basic monsters (Rockitten). But we do this, then we can also give 6 techniques to standalone monsters (the ones with no evolutions)! Everything in another PR. @Sanglorian  what do you think?

Sanglorian commented:
Around 6 techniques for stand alone monsters makes sense -- but really I'd give them as many as makes sense. It's nice to have choice, so at least 5 makes sense but it could be 6 or 9 or 12 for an individual species.

JaskRendix commented:
Ok, updated the last moves, ready to merge. In another PR I'll deal with the monster without evolution.

bitcraft commented on tuxemon/db.py:
```diff
@@ -233,6 +233,13 @@ def valid_level(cls, v):
+        if v > 0:
+            if has.multiple(v, 2):
+                return v
+            raise ValueError(f"no resource exists in db: {v}")
```
this error message is misleading

JaskRendix replied:
I'm going to remove it, I used it for maintaining order

bitcraft commented on tuxemon/db.py:
```diff
@@ -254,6 +261,13 @@ class MonsterEvolutionItemModel(BaseModel):
+        if v > 0:
+            if has.multiple(v, 3):
+                return v
+            raise ValueError(f"no resource exists in db: {v}")
```
this error message is misleading

JaskRendix replied:
I'm going to remove it, I used it for maintaining order

bitcraft commented on tuxemon/db.py:
```diff
@@ -1019,6 +1033,21 @@ def db_entry(self, table: TableName, slug: str) -> bool:
             return True
         return False
 
+    def multiple(self, num: int, check_with: int) -> bool:
```
are you sure we need a function that just does `X % Y == 0`?

JaskRendix replied:
I'm going to remove it, I used it for maintaining order

JaskRendix committed:
monster moves
